### PR TITLE
Dry run

### PR DESCRIPTION
**Pull request by facuMH**
The idea behind this is to able to simulate a run to analyze behavior and how would celerity scale to larger amounts nodes, without necessarily having those nodes.
For this we are not flushing commands unless they are epochs ( and this ones only to the master node).

### CODE REVIEW

File: src/runtime.cc [hunk 1]
```diff
@@ -102,13 +102,18 @@ namespace detail {
 		MPI_Comm_size(MPI_COMM_WORLD, &world_size);
 		num_nodes = world_size;
 
+		cfg = std::make_unique<config>(argc, argv);
+		if(is_dry_run()) {
+			if(num_nodes != 1) throw std::runtime_error("YOU CAN NOT MAKE A DRY RUN WITH MPI.\n");
```
Comment (psalz) [inline src/runtime.cc:107]:
```suggestion
			if(num_nodes != 1) throw std::runtime_error("Cannot perform a dry run with more than one rank.\n");
```

File: src/config.cc [hunk 2]
```diff
@@ -180,6 +180,16 @@ namespace detail {
 			const auto result = get_env("CELERITY_FORCE_WG");
 			if(result.first) { CELERITY_WARN("Support for CELERITY_FORCE_WG has been removed with Celerity 0.3.0."); }
 		}
+
+		// -------------------------------- CELERITY_DRY_RUN ---------------------------------
+		{
+			const auto result = get_env("CELERITY_DRYRUN_NODES");
+			if(result.first) {
+				dry_run = std::stoi(result.second) > 0;
```
Comment (psalz) [inline src/config.cc:188]: You can use the `parse_uint` helper here to make this more robust!

File: src/config.cc [hunk 3]
```diff
@@ -180,6 +180,16 @@ namespace detail {
...
+				dry_run_nodes = std::stoi(result.second);
+				CELERITY_WARN("Celerity Dry Run detected.");
```
Comment (psalz) [inline src/config.cc:190]: As this is not a validation warning/error, I would suggest moving it to the runtime constructor! Also maybe rephrase to `"Performing a dry run with {} simulated nodes."`.

File: include/config.h [hunk 4]
```diff
@@ -40,6 +40,8 @@ namespace detail {
 		 */
 		const std::optional<device_config>& get_device_config() const { return device_cfg; };
 		std::optional<bool> get_enable_device_profiling() const { return enable_device_profiling; };
+		std::optional<bool> is_dry_run() const { return dry_run; };
```
Comment (psalz) [inline include/config.h:43]: I think we could just default to `false` if it is not set and void the `optional`.

File: include/runtime.h [hunk 5]
```diff
@@ -77,6 +77,10 @@ namespace detail {
 
 		host_object_manager& get_host_object_manager() const;
 
+		bool is_dry_run() const { return cfg->is_dry_run().value_or(false); }
+
+		int get_command_count() const;
```
Comment (psalz) [inline include/runtime.h:82]: I'd prefer not introducing to new members to the runtime's API for a single test case. Why not use `runtime_testspy` instead?

File: test/runtime_tests.cc [hunk 6]
```diff
@@ -992,5 +992,39 @@ namespace detail {
 		});
 	}
 
+	void dry_run_with_nodes(const int nodes) {
+		const std::string dryrun_envvar_name = "CELERITY_DRYRUN_NODES";
+		setenv(dryrun_envvar_name.c_str(), std::to_string(nodes).c_str(), 1);
```
Comment (psalz) [inline test/runtime_tests.cc:997]: For the device selection tests we've used a `config_testspy` instead of modifying the environment, imo that's a cleaner approach.

File: test/runtime_tests.cc [hunk 7]
```diff
@@ -992,5 +992,39 @@ namespace detail {
...
+		test_utils::maybe_print_graph(tm);
+	}
+
+	TEST_CASE_METHOD(test_utils::runtime_fixture, "dry run", "[dryrun]") {
+		SECTION("with 2 nodes") { dry_run_with_nodes(2); }
+		SECTION("with 8 nodes") { dry_run_with_nodes(8); }
+		SECTION("with 16 nodes") { dry_run_with_nodes(16); }
+		SECTION("with 32 nodes") { dry_run_with_nodes(32); }
+		SECTION("with 64 nodes") { dry_run_with_nodes(64); }
+		SECTION("with 128 nodes") { dry_run_with_nodes(128); }
```
Comment (psalz) [inline test/runtime_tests.cc:1025]: Use a generator instead? Also, maybe 2 or 3 configs are sufficient?

File: test/runtime_tests.cc [hunk 8]
```diff
@@ -992,5 +992,39 @@ namespace detail {
...
+			cgh.parallel_for<class UKN(dryrun)>(buff.get_range(), [=](item<1> item) { acc[item.get_id()] = item.get_id().get(0); });
+		});
+		q.slow_full_sync();
+
+		// (intial epoch + task + sync epoch) per node.
+		CHECK(rt.get_command_count() == 3 * nodes);
+		test_utils::maybe_print_graph(tm);
+	}
+
+	TEST_CASE_METHOD(test_utils::runtime_fixture, "dry run", "[dryrun]") {
```
Comment (psalz) [inline test/runtime_tests.cc:1019]:
```suggestion
	TEST_CASE_METHOD(test_utils::runtime_fixture, "Dry runs generate graphs for an arbitrary number of simulated worker nodes", "[dryrun]") {
```

File: test/runtime_tests.cc [hunk 9]
```diff
@@ -992,5 +992,39 @@ namespace detail {
...
+		auto& tm = rt.get_task_manager();
+		tm.set_horizon_step(2);
+
+		CHECK(rt.is_dry_run());
+
+		buffer<size_t, 1> buff{nodes * 2};
+		q.submit([=](handler& cgh) {
+			accessor acc{buff, cgh, celerity::access::one_to_one{}, write_only};
+			cgh.parallel_for<class UKN(dryrun)>(buff.get_range(), [=](item<1> item) { acc[item.get_id()] = item.get_id().get(0); });
+		});
```
Comment (psalz) [inline test/runtime_tests.cc:1011]:
```suggestion
		q.submit([=](handler& cgh) {
			cgh.host_task(range<1>{nodes * 2}, [] {});
		});
```

File: src/config.cc [hunk 10]
```diff
@@ -180,6 +180,16 @@ namespace detail {
 			const auto result = get_env("CELERITY_FORCE_WG");
 			if(result.first) { CELERITY_WARN("Support for CELERITY_FORCE_WG has been removed with Celerity 0.3.0."); }
 		}
+
+		// -------------------------------- CELERITY_DRY_RUN ---------------------------------
```
Comment (psalz) [inline src/config.cc:184]:
```suggestion
		// -------------------------------- CELERITY_DRYRUN_NODES ---------------------------------
```

**Review by psalz — CHANGES_REQUESTED** [review]
Thanks! I've added some notes. Also, I'm wondering whether we should call the variable `CELERITY_DRY_RUN_NODES` instead? 🤔

File: include/config.h (hunk 4 above)
Comment (PeterTh) [inline include/config.h:43]: I don't see why this needs to be a `std::optional<bool>`, when it could just be a `bool`. I don't think we'll ever need to distinguish "dry run not set" from "not a dry run".

File: test/runtime_tests.cc (hunk 6 above)
Comment (PeterTh) [reply test/runtime_tests.cc:997]: I feel like using setenv in the unit tests is fine. It's slightly more integration- than unit-testy, but eh.

**Review by PeterTh — COMMENTED** [review]
Looks pretty good, as long as you take care of the comments. Also, DON'T SCREAM AT THE USER :P

File: test/runtime_tests.cc (hunk 6 above)
Comment (facuMH) [reply test/runtime_tests.cc:997]: not sure if we can do the same here, because in `device_selection_tests` the `config_testspy` is modifying a hand-made config, not the config that is initialized inside the runtime, which is what one of the things we want to test here.

Comment (psalz) [reply test/runtime_tests.cc:997]: Right, because it's setting `num_nodes` in the constructor. In any case, modifying the global environment from a test is always iffy. What happens when the test fails? You'd need at least a RAII wrapper to ensure that the environment is always restored.

But maybe the cleanest solution would be to change the runtime ctor, like so:
```diff
-	runtime::runtime(int* argc, char** argv[], device_or_selector user_device_or_selector) {
+	runtime::runtime(int* argc, char** argv[], config cfg = {argc, argv}, device_or_selector user_device_or_selector) {
```

File: test/test_utils.h [hunk 11]
```diff
@@ -427,6 +427,17 @@ namespace test_utils {
 		}
 	}
 
+	struct set_test_env {
+		set_test_env(const std::string& evn, const std::string val) {
+			setenv(evn.c_str(), val.c_str(), 1);
```
Comment (psalz) [inline test/test_utils.h:432]: It looks like `setenv` does not exist on Windows, so this is not portable unfortunately. Also if you want this to be a generic utility (which I assume as it is in `test_utils.h`) I think a more robust solution would be to restore the previous value upon destruction.

**Review by psalz — CHANGES_REQUESTED** [review]

File: test/test_utils.h [hunk 12]
```diff
@@ -427,6 +427,17 @@ namespace test_utils {
 		}
 	}
 
+	struct set_test_env {
+		set_test_env(const std::string& evn, const std::string val) {
```
Comment (psalz) [inline test/test_utils.h:431]:
```suggestion
		set_test_env(const std::string& env, const std::string val) {
```

File: test/runtime_tests.cc [hunk 13]
```diff
@@ -992,5 +995,30 @@ namespace detail {
...
+
+		q.submit([=](handler& cgh) { cgh.host_task(range<1>{nodes * 2}, [](partition<1>) {}); });
+		q.slow_full_sync();
+
+		// (intial epoch + task + sync epoch) per node.
+		CHECK(runtime_testspy::get_command_count(rt) == 3 * nodes);
+		test_utils::maybe_print_graph(tm);
+	}
+
+	TEST_CASE_METHOD(test_utils::runtime_fixture, "Dry run generate grap", "[dryrun]") {
```
Comment (psalz) [inline test/runtime_tests.cc:1018]: Something went wrong with the name change here!

File: src/runtime.cc [hunk 14]
```diff
@@ -102,13 +102,19 @@ namespace detail {
 		MPI_Comm_size(MPI_COMM_WORLD, &world_size);
 		num_nodes = world_size;
 
+		cfg = std::make_unique<config>(argc, argv);
+		if(is_dry_run()) {
```
Comment (psalz) [inline src/runtime.cc:106]:
```suggestion
		if(cfg->is_dry_run()) {
```

A minor thing, but otherwise it's not clear that config must exist before `runtime::is_dry_run` can be called.

File: test/runtime_tests.cc [hunk 15]
```diff
@@ -992,5 +995,30 @@ namespace detail {
...
+		q.slow_full_sync();
+
+		// (intial epoch + task + sync epoch) per node.
+		CHECK(runtime_testspy::get_command_count(rt) == 3 * nodes);
+		test_utils::maybe_print_graph(tm);
+	}
+
+	TEST_CASE_METHOD(test_utils::runtime_fixture, "Dry run generates commands for an arbitrary number of simulated worker nodes", "[dryrun]") {
+		const size_t nodes = GENERATE(values({4, 8, 16}));
+		SECTION("with {} nodes", std::to_string(nodes).c_str()) { dry_run_with_nodes(nodes); }
```
Comment (psalz) [inline test/runtime_tests.cc:1020]:
```suggestion
		SECTION("with " << nodes << " nodes", dry_run_with_nodes(nodes); }
```

Catch2 does not support fmt-style formatting.

File: test/test_utils.h [hunk 16]
```diff
@@ -427,6 +432,50 @@ namespace test_utils {
...
+		set_test_env(const std::string& env, const std::string val) {
+			name_size = GetEnvironmentVariableA(env.c_str(), nullptr, 0);
+			if(name_size > 0) {
+				original_value.resize(name_size);
+				GetEnvironmentVariableA(env.c_str(), &original_value[0], name_size + 1);
+			}
+			SetEnvironmentVariable(env.c_str(), val.c_str());
+			env_var_name = env;
+		}
+		// if it was not previously set, then original_value is 0. using setEnvVar with 0 deletes the env var.
```
Comment (psalz) [inline test/test_utils.h:452]: This comment seems outdated?

File: test/test_utils.h [hunk 17]
```diff
@@ -427,6 +432,50 @@ namespace test_utils {
...
+		~set_test_env() {
+			if(original_value.size()) {
```
Comment (psalz) [inline test/test_utils.h:454]:
```suggestion
			if(!original_value.empty()) {
```

File: test/test_utils.h [hunk 18]
```diff
@@ -427,6 +432,50 @@ namespace test_utils {
...
+
+	  public:
+		set_test_env(const std::string& env, const std::string val) {
+			const char* has_value = std::getenv(env.c_str());
+			if(has_value != nullptr) { original_value = has_value; }
+			setenv(env.c_str(), val.c_str(), 1);
+			env_var_name = env;
+		}
+		~set_test_env() {
+			if(!original_value.size()) {
```
Comment (psalz) [inline test/test_utils.h:470]: The branches here are reversed, which is somewhat confusing.

**Review by psalz — CHANGES_REQUESTED** [review]
I've added a few more notes. Also I vote in favor of renaming the env var to `CELERITY_DRY_RUN_NODES` - thoughts?

**Review by BlackMark29A — CHANGES_REQUESTED** [review]
I've identified some spots where readability/documentation could be improvied, but there are also some bugs which crept in:

**Review by BlackMark29A — CHANGES_REQUESTED** [review]

**Review by BlackMark29A — APPROVED** [review]
Thank you. Looks good to me now.

File: src/runtime.cc [hunk 19]
```diff
@@ -102,13 +102,19 @@ namespace detail {
 		MPI_Comm_size(MPI_COMM_WORLD, &world_size);
 		m_num_nodes = world_size;
 
+		m_cfg = std::make_unique<config>(argc, argv);
+		if(m_cfg->is_dry_run()) {
+			if(m_num_nodes != 1) throw std::runtime_error("Cannot perform a dry run with more than one rank.\n");
```
Comment (fknorr) [inline src/runtime.cc:107]: That error message might be confusing, it can be read as "The only allowed parameter value for CELERITY_DRY_RUN_NODES is 1". Maybe go with "Celerity must be launched with a single MPI process when specifying CELERITY_DRY_RUN_NODES" or similar.

**Review by fknorr — APPROVED** [review]
LGTM up to the error message, thanks!

**Review by psalz — APPROVED** [review]
LGTM!
